pkcs11session.c

```diff
@@ -79,23 +100,141 @@ ZEND_BEGIN_ARG_INFO_EX(arginfo_destroyObject, 0, 0, 1)
     ZEND_ARG_INFO(0, object)
 ZEND_END_ARG_INFO()
 
-PHP_METHOD(Session, getInfo) {
+extern zend_class_entry *ce_Pkcs11_Module;
+PHP_METHOD(Session, __construct) {
+    pkcs11_session_object *objval = Z_PKCS11_SESSION_P(ZEND_THIS);
+    objval->session = 0; /* not initialized yet */
 
+    CK_SLOT_ID slotid;
+    CK_FLAGS flags;
     CK_RV rv;
-    CK_SESSION_INFO sessionInfo;
 
+    zval *php_pkcs11;
+    zend_long php_slotid = 0;
+    zend_long php_flags = 0;
+
+    ZEND_PARSE_PARAMETERS_START(3, 3)
+      Z_PARAM_OBJECT_OF_CLASS(php_pkcs11, ce_Pkcs11_Module)
+      Z_PARAM_LONG(php_slotid)
+      Z_PARAM_LONG(php_flags)
+    ZEND_PARSE_PARAMETERS_END();
+
+    objval->pkcs11 = Z_PKCS11_P(php_pkcs11);
+    slotid = (CK_SLOT_ID)php_slotid;
+    flags = (CK_FLAGS)php_flags;
+
+    if (flags &
+        (CKF_RW_SESSION || CKF_SERIAL_SESSION)) {
+      ; /* nope */
+    } else {
+      ; /* nope */
+    }
+
+    CK_SESSION_HANDLE hSession;
+    if (ZEND_NUM_ARGS() > 4)
+        rv = objval->pkcs11->functionList->C_OpenSession(slotid, flags, NULL, NULL, &hSession); /* TODO: add callbacks */
+    else
+        rv = objval->pkcs11->functionList->C_OpenSession(slotid, flags, NULL, NULL, &hSession); /* TODO: add callbacks */
+
+    if (rv != CKR_OK) {
+          pkcs11_error(rv, "Unable to instanciate a session");
+          return;
+    }
+    objval->session = hSession;
+}
+
+PHP_METHOD(Session, __destruct) {
     pkcs11_session_object *objval = Z_PKCS11_SESSION_P(ZEND_THIS);
+
+    CK_RV rv;
+
+    ZEND_PARSE_PARAMETERS_NONE();
+
+    rv = objval->pkcs11->functionList->C_CloseSession(objval->session);
+    if (rv != CKR_OK)
+        pkcs11_error(rv, "could not be C_CloseSession()'d");
+}
```

Comment:
C_CloseSession is already called in pkcs11_session_shutdown(). This is redundant.

Reply:
you are right: Same as the previous comment, I missed the magics from DEFINE_MAGIC_FUNCS() / DECLARE_MAGIC_FUNCS().

Again, what's about adding if (rv != CKR_OK) pkcs11_error(rv, "could not be C_CloseSession()'d"); ?